NAV-917: conda constrained for old navigator releases

yshmatov-anaconda opened the pull request:
Fixed issues:
- `anaconda-navigator<2.4.0` doesn't completely support `conda>=23.1.0`
- `anaconda-navigator<2.3.0` doesn't completely support `pyqt>=5.15,<6.0a0`
- `anaconda-navigator` has issues with `conda>=22.11.0,<23.1.0`
- `anaconda-client<1.10.0` is incompatible with python 3.10

intentionally-left-nil reviewed (commented):
Can we add unit tests to verify this has the correct behavior?

yshmatov-anaconda commented on main.py:
```diff
@@ -1184,11 +1194,64 @@ def patch_record_in_place(fn, record, subdir):
+        old = [old]
+
+    for item in old:
+        try:
```
Both options should work similarly in the end. But I see no reason to look for the item twice (`in` and then `index`).

chenghlee replied:
Trying to avoid two lookups (`in` + `index`) smells a bit like premature optimization to me. IMO, the new implementation is harder to read, and it's not at all clear whether it preserves the in-place replacement behavior of the previous implementation.

(Also: the stack unwinding from `try`/`except` is going to be far more expensive than an index lookup.)

yshmatov-anaconda replied:
If majority wants to have `in` + `index` - no problem. It's, in any case, just another way to write the same thing. And I didn't even think that we actually require high level of performance for this script.

Yes - it preserves. And I also tried to put all primary details in the docstring for the function. But, in other words - it takes list from the first argument, and in it replaces all values from the second argument with the third argument. (_P.S. second argument might be not only a single string value, but also a list of values_). And it makes sure that we end up having third value mentioned only once in the output if we are doing a replacement. In case we need to make sure that we end up with a new value in the list even if we replaced nothing - `append=True` is the way to do it.

There are also separate tests for this particular function where anyone should be able to find different use cases and see how they work.

But, anyways - I checked multiple times this updated function - there are no unexpected changes in the `repodata-patched.json` at all, only anaconda-navigator and anaconda-client related ones (_checked for multiple primary platforms_). Which definitely means that it doesn't break previous behavior that other developers before me expected and used.

yshmatov-anaconda commented:
> Can we add unit tests to verify this has the correct behavior?

What do you want to check? If you are talking only about the updated `replace_dep` - then yes. But I also checked it when I created it to make sure it has no negative effects on what we already had.

If you are talking about the change to the repodata - I also checked the output of this change and it ended up with expected changes to all repodata files. Since we already know how it should affect everything - I believe it should be enough. Unless we want to double-check for repeated uses (I only designed the change with this in mind, but did not check this behavior; But it might also be checked right after this change is applied for the first time).

yshmatov-anaconda commented:
I decided not to wait for answers and made few changes:

- added tests, but also changed the behavior of the `replace_dep` function a little bit. If you provide multiple items to replace/remove - it would now affect all matching items instead of only the first one. This should not affect how the function is used currently, but otherwise it should help getting cleaner outputs in case data is already messed up.
- rebased this PR to the latest `master` and rechecked that everything is still fine.
- looked into adding tests for idempotency check, but only half-the way discovered that it is unnecessary since we always update the "raw" repodata instead of the "previously updated". However - I still found few places that required some cleanup. This change is not included in this PR, but if you are interested you may find it in this commit: [ed86cb00cf7d925a2bb4cd1306999fb03b60ecde](https://github.com/yshmatov-anaconda/repodata-hotfixes/commit/ed86cb00cf7d925a2bb4cd1306999fb03b60ecde).

yshmatov-anaconda commented on main.py:
```diff
@@ -848,6 +851,13 @@ def patch_record_in_place(fn, record, subdir):
+        if re.match(r'1\.|2\.[0-2]\.', version):  # < 2.3.0
+            replace_dep(depends, ['pyqt >=5.6,<6.0a0', 'pyqt >=5.6', 'pyqt'], 'pyqt >=5.6,<5.15')
+
+        if re.match(r'1\.|2\.[0-3]\.', version):  # < 2.4.0
+            replace_dep(depends, 'conda', 'conda <22.11.0', append=True)
```
`append=True` will just add `conda <22.11.0` record if it is not there yet. I mentioned this behavior in the docstring for the updated `replace_dep`.

And yes - old versions don't have `conda` in the dependencies explicitly at all, only extra stuff like `conda-repo-cli` and `conda-token`.
`append=True` just ensures that we would have `conda < 22.11.0` for all old releases in any case. Which is, in my opinion, a valuable thing to have. Otherwise there might be cases when `conda install conda>=23.1.0 anaconda-navigator` may end up resolving it with pretty old navigator release, which just happens to have no conda dependency at all.

yshmatov-anaconda commented:
Changed `try-except` with `if-in-then` as was requested.

And also checked again the output. For `linux-64`, `noarch`, `osx-64`, `osx-arm64`, `win-64` - there are only expected changes mentioned in the description to this PR.
